File: 05-Algorithms/code/typescript/src/bridges.ts
import { buildAdjList, type Edge } from "./graphCommon.js";

/**
 * Finds every bridge (an edge whose removal disconnects the graph) in an
 * undirected graph on n vertices, via the standard discovery-time / low-link
 * DFS. Implemented iteratively to avoid recursion-depth limits. Each bridge
 * is returned as `[min(u,v), max(u,v)]`, and the overall list is sorted
 * ascending, so the output is independent of DFS traversal order and of
 * input edge order. Parallel edges between the same pair are never
 * bridges (removing one still leaves the other), which this implementation
 * accounts for by tracking the edge index used to enter each vertex rather
 * than just the parent vertex. O(V + E).
 */
export function bridges(n: number, edges: Edge[]): Edge[] {
  // Build an adjacency list of (neighbor, edgeId) pairs so parallel edges
  // are distinguishable from one another.
  const adj: [number, number][][] = Array.from({ length: n }, () => []);
  edges.forEach(([u, v], edgeId) => {
    adj[u].push([v, edgeId]);
    adj[v].push([u, edgeId]);
  });
  for (const list of adj) list.sort((a, b) => a[0] - b[0]);

  const disc = new Array(n).fill(-1);
  const low = new Array(n).fill(-1);
  let timer = 0;
  const bridgeEdgeIds = new Set<number>();

  for (let start = 0; start < n; start++) {
    if (disc[start] !== -1) continue;
    const stack: { u: number; parentEdgeId: number; i: number }[] = [{ u: start, parentEdgeId: -1, i: 0 }];
    disc[start] = low[start] = timer++;
    while (stack.length > 0) {
      const frame = stack[stack.length - 1];
      const neighbors = adj[frame.u];
      if (frame.i < neighbors.length) {
        const [v, edgeId] = neighbors[frame.i++];
        if (edgeId === frame.parentEdgeId) continue;
        if (disc[v] === -1) {
          disc[v] = low[v] = timer++;
          stack.push({ u: v, parentEdgeId: edgeId, i: 0 });
        } else {
          low[frame.u] = Math.min(low[frame.u], disc[v]);
        }
      } else {
        stack.pop();
        if (stack.length > 0) {
          const parentFrame = stack[stack.length - 1];
          low[parentFrame.u] = Math.min(low[parentFrame.u], low[frame.u]);
          if (low[frame.u] > disc[parentFrame.u]) bridgeEdgeIds.add(frame.parentEdgeId);
        }
      }
    }
  }

  const result: Edge[] = [];
  edges.forEach(([u, v], edgeId) => {
    if (bridgeEdgeIds.has(edgeId)) result.push(u < v ? [u, v] : [v, u]);
  });
  result.sort((a, b) => (a[0] !== b[0] ? a[0] - b[0] : a[1] - b[1]));
  return result;
}

File: 05-Algorithms/code/python/seb_alg/bridges.py
"""Mirrors `typescript/src/bridges.ts` exactly."""

from __future__ import annotations

from typing import List, Sequence, Tuple


def bridges(n: int, edges: Sequence[Sequence[int]]) -> List[List[int]]:
    """Finds every bridge in an undirected graph via discovery-time /
    low-link DFS (iterative). Each bridge is [min(u,v), max(u,v)], overall
    list sorted ascending. Parallel edges are never bridges, tracked via
    edge index (not just parent vertex) to distinguish them. O(V + E).
    """
    adj: List[List[Tuple[int, int]]] = [[] for _ in range(n)]
    for edge_id, (u, v) in enumerate(edges):
        adj[u].append((v, edge_id))
        adj[v].append((u, edge_id))
    for lst in adj:
        lst.sort(key=lambda pair: pair[0])

    disc = [-1] * n
    low = [-1] * n
    timer = 0
    bridge_edge_ids = set()

    for start in range(n):
        if disc[start] != -1:
            continue
        stack = [[start, -1, 0]]
        disc[start] = low[start] = timer
        timer += 1
        while stack:
            frame = stack[-1]
            u, parent_edge_id, i = frame[0], frame[1], frame[2]
            neighbors = adj[u]
            if i < len(neighbors):
                v, edge_id = neighbors[i]
                frame[2] += 1
                if edge_id == parent_edge_id:
                    continue
                if disc[v] == -1:
                    disc[v] = low[v] = timer
                    timer += 1
                    stack.append([v, edge_id, 0])
                else:
                    low[u] = min(low[u], disc[v])
            else:
                stack.pop()
                if stack:
                    parent_frame = stack[-1]
                    pu = parent_frame[0]
                    low[pu] = min(low[pu], low[u])
                    if low[u] > disc[pu]:
                        bridge_edge_ids.add(parent_edge_id)

    result: List[List[int]] = []
    for edge_id, (u, v) in enumerate(edges):
        if edge_id in bridge_edge_ids:
            result.append([u, v] if u < v else [v, u])
    result.sort(key=lambda e: (e[0], e[1]))
    return result
